ios/keyman/Keyman/Keyman/AppDelegate.swift

```diff
@@ -91,6 +113,74 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     return _overlayWindow!
   }
 
+  public func installAdhocKeyboard(url: URL) {
+    let documentsDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
+    var destination =  documentsDirectory
+    destination.appendPathComponent("temp/\(url.lastPathComponent)")
+
+    Manager.shared.unzipFile(fileUrl: url, destination: destination) {
+      self.promptAdHocInstall(destination)
+    }
+  }
+
+  public func promptAdHocInstall(_ folder: URL) {
+    _adhocDirectory = folder
+
+    var welcomePath = folder
+    welcomePath.appendPathComponent("Welcome.htm")
+
+    let vc = UIViewController()
+    vc.view.backgroundColor = .red
+    let wkWebView = WKWebView.init(frame: vc.view.frame)
+    wkWebView.backgroundColor = .blue
+    vc.view.addSubview(wkWebView)
+    let cancelBtn = UIBarButtonItem(title: "Cancel", style: .plain,
+                                    target: self,
+                                    action: #selector(cancelAdHocBtnHandler))
+    let installBtn = UIBarButtonItem(title: "Install", style: .plain,
+                                     target: self,
+                                     action: #selector(installAdHocBtnHandler))
+    vc.navigationItem.leftBarButtonItem = cancelBtn
+    vc.navigationItem.rightBarButtonItem = installBtn
+    let nvc = UINavigationController.init(rootViewController: vc)
+
+    self.window?.rootViewController?.present(nvc, animated: true, completion: {
+      if FileManager.default.fileExists(atPath: welcomePath.path) {
+        print("loading request: \(welcomePath)")
+        if let html = try? String(contentsOfFile: welcomePath.path, encoding: String.Encoding.utf8) {
+          wkWebView.loadHTMLString(html, baseURL: nil)
+        } else {
+           wkWebView.loadHTMLString("Keyboard Package", baseURL: nil)
+        }
+      } else {
+        wkWebView.loadHTMLString("Keyboard Package", baseURL: nil)
+      }
+    })
+  }
+
+  @objc func installAdHocBtnHandler() {
+    if let adhocDir = _adhocDirectory {
+      Manager.shared.parseKMP(adhocDir, complete: { count in
+        print("count: \(count)")
+        self.window?.rootViewController?.dismiss(animated: true, completion: {
+          let title = count == 0 ? "Error" : "Success"
+          let message = "\(count) keyboard(s) installed"
+          let alertController = UIAlertController(title: title, message:
+           message, preferredStyle: UIAlertControllerStyle.alert)
+          alertController.addAction(UIAlertAction(title: "OK",
+                                                  style: UIAlertActionStyle.default,
+                                                  handler: nil))
+
+           self.window?.rootViewController?.present(alertController, animated: true, completion: nil)
+        })
+      })
+    }
+  }
+
+  @objc func cancelAdHocBtnHandler() {
+    self.window?.rootViewController?.dismiss(animated: true, completion: nil)
```

Comment:
Do you need to do any cleanup (deleting temporary zip files / extract folders)?

Reply:
Yes, I am adding that tonight. As well as a better method for handling all things KMP.